python/ray/train/lightning/lightning_trainer.py

```diff
@@ -519,9 +527,19 @@ def _lightning_train_loop_per_worker(config):
         logger.warning(
             "`strategy` specified in `LightningConfig.trainer_init_config` "
             "will be ignored. LightningTrainer will create a RayDDPStrategy "
-            "object based on `LightningConfig.ddp_strategy_config`."
+            "or RayFSDPStrategy object based on `LightningConfigBuilder.strategy()`."
         )
-    trainer_config["strategy"] = RayDDPStrategy(**ddp_strategy_config)
+
+    assert strategy_name in [
+        "ddp",
+        "fsdp",
+    ], "LightningTrainer currently supports 'ddp' and 'fsdp' strategy. "
+    "Please choose one of them."
+
+    if strategy_name == "ddp":
+        trainer_config["strategy"] = RayDDPStrategy(**strategy_config)
+    if strategy_name == "fsdp":
+        trainer_config["strategy"] = RayFSDPStrategy(**strategy_config)
```

Comment:
stupid question. should user simply specify the Strategy object? or do we usually use a string name?

Reply:
I agree that it'd be more natural to pass a Strategy object instead of given a list of initialization kwargs. The problem is that we have to modify two class methods of the Strategy class in order to be compatible with the Ray environment.

Let me search to see if there's a way to modify class methods given an object.

Reply:
Previously, we subclassed `DDPStrategy` and `FSDPStrategy`, and initialized a strategy instance with a list of init kwargs provided by the users. 

I found that we can dynamically change property methods of an instance with `setattr`, so that the users only need to provide a strategy object, instead of an argument list. We change these two property methods on our end.

```python
from types import MethodType

@property
def root_device(self) -> torch.device:
    return ray.train.torch.get_device()

@property
def distributed_sampler_kwargs(self) -> Dict[str, Any]:
    return dict(
        num_replicas=self.world_size,
        rank=self.global_rank,
    )

# In LightningTrainer's worker loop

# - Get the strategy object
strategy = lightning_config["_trainer_config"].get("strategy", DDPStrategy())

# - Dynamically change properties to our implementations
setattr(strategy.__class__, "root_device", root_device)
setattr(strategy.__class__, "distributed_sampler_kwargs", distributed_sampler_kwargs)
```

 @gjoliver Do you think it's a good way?

Reply:
IMO, I prefer just passing the string over setattr.

Reply:
oh, the problem is when users construct the Strategy object, they wouldn't be able to know the global_rank?
I imagine ray.train.torch.get_device() would do the right thing no matter you set it or not?
we probably shouldn't change these 2 properties to class methods, feels a bit hacky.
thanks for the investigation.